### PR TITLE
Ash runner: fix crash on resize/minimize and add shader hot reloading

### PR DESCRIPTION
**Pull request by Hentropy**
- rebuild the swapchain when the window is resized, pipelines use dynamic state so they are not rebuilt.
- F5 will rebuild the shader crate, load the spirv, and rebuild the pipeline all on a separate thread.

I also moved the ash setup code around quite a bit, so it looks like a big change when in reality it's like ~75% the same code.

### CODE REVIEW

File: examples/runners/ash/src/main.rs
Comment (khyperia) [inline examples/runners/ash/src/main.rs:126]: this won't work on mac/linux. instead:
```rust
let spirv_codegen_backend = format!(
    "codegen_backend={}rustc_codegen_spirv{}",
    std::env::consts::DLL_PREFIX,
    std::env::consts::DLL_SUFFIX
);
```

File: examples/runners/ash/src/main.rs
Comment (khyperia) [inline examples/runners/ash/src/main.rs:129]: same here, windows slashes

**Review by khyperia — COMMENTED** [review]
Looks good! I'd like someone else to take a look at it since I don't know `ash` very well, but seems good to me.

File: examples/runners/ash/src/main.rs
Comment (Jasper-Bekkers) [inline examples/runners/ash/src/main.rs:46]: While the logic you wrote below seems like it could probably work correctly in this case; I think this kind of state management through atomic booleans is ripe for potential race conditions down the line. Would it be an option to re-do this part of the logic to use an mpsc channel instead to keep it a bit more tidy and less likely to have threading bugs?

File: examples/runners/ash/src/main.rs
Comment (Jasper-Bekkers) [inline examples/runners/ash/src/main.rs:220]: It would be nice if somebody could run and test this on macOS before we merge it, mostly would like to see the resize code get tested there.

File: examples/runners/ash/src/main.rs
Comment (Jasper-Bekkers) [inline examples/runners/ash/src/main.rs:116]: It looks like you're not resizing the swapchain (at least not passing through the new window resolution here)?

There should probably also be a condition around resizing down to a zero sized swapchain, this can happen at least on Windows when minimizing the window, but also when resizing the window to be really tiny.

**Review by Jasper-Bekkers — COMMENTED** [review]
Running with --debug-layer results in quite a few validation error messages on shutdown (I think it gets worse with more reloads, but I'm not 100% sure), I think those should get fixed before we merge this in.

```cmd
ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xfe25f60000000013, type = VK_OBJECT_TYPE_SHADER_MODULE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkShaderModule 0xfe25f60000000013[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0x51b3510000000014, type = VK_OBJECT_TYPE_PIPELINE_LAYOUT; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkPipelineLayout 0x51b3510000000014[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0x6269300000000015, type = VK_OBJECT_TYPE_PIPELINE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkPipeline 0x6269300000000015[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xb5f68b000000000e, type = VK_OBJECT_TYPE_RENDER_PASS; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkRenderPass 0xb5f68b000000000e[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0x1f91b40000000011, type = VK_OBJECT_TYPE_FRAMEBUFFER; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkFramebuffer 0x1f91b40000000011[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xc6ac6a000000000f, type = VK_OBJECT_TYPE_FRAMEBUFFER; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkFramebuffer 0xc6ac6a000000000f[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

ERROR:
VALIDATION [VUID-vkDestroyDevice-device-00378 (1901072314)] : Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x168e2411ca8, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xedbd50000000010, type = VK_OBJECT_TYPE_FRAMEBUFFER; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x168e2411ca8[], VkFramebuffer 0xedbd50000000010[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)
```

I don't know if you intended to fix minimizing the window with this PR however I think it would be a small (but nice) addition to this. Right now we're getting validation layer messages around this as well.

```cmd
ERROR:
VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01689 (320081257)] : Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01689 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x13140d69 | vkCreateSwapchainKHR(): pCreateInfo->imageExtent = (0, 0) which is illegal. The Vulkan spec states: imageExtent members width and height must both be non-zero (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01689)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-width-00885 (-1231547098)] : Validation Error: [ VUID-VkFramebufferCreateInfo-width-00885 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xb6981526 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo width must be greater than zero. The Vulkan spec states: width must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-width-00885)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-height-00887 (-1513456701)] : Validation Error: [ VUID-VkFramebufferCreateInfo-height-00887 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa5ca7bc3 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo height must be greater than zero. The Vulkan spec states: height must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-height-00887)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-width-00885 (-1231547098)] : Validation Error: [ VUID-VkFramebufferCreateInfo-width-00885 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xb6981526 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo width must be greater than zero. The Vulkan spec states: width must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-width-00885)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-height-00887 (-1513456701)] : Validation Error: [ VUID-VkFramebufferCreateInfo-height-00887 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa5ca7bc3 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo height must be greater than zero. The Vulkan spec states: height must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-height-00887)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-width-00885 (-1231547098)] : Validation Error: [ VUID-VkFramebufferCreateInfo-width-00885 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xb6981526 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo width must be greater than zero. The Vulkan spec states: width must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-width-00885)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-height-00887 (-1513456701)] : Validation Error: [ VUID-VkFramebufferCreateInfo-height-00887 ] Object 0: handle = 0x1b041ce94f8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xa5ca7bc3 | vkCreateFramebuffer(): Requested VkFramebufferCreateInfo height must be greater than zero. The Vulkan spec states: height must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.2.148.0/windows/1.2-extensions/vkspec.html#VUID-VkFramebufferCreateInfo-height-00887)
```

File: examples/runners/ash/src/main.rs
Comment (Hentropy) [reply examples/runners/ash/src/main.rs:212]: Linux should get checked again as well.

File: examples/runners/ash/src/main.rs
Comment (Hentropy) [reply examples/runners/ash/src/main.rs:108]: Both issues here have been addressed.  Viewports and scissors are now recreated when the window is resized, and rendering is paused when width or height is 0.  Resolution still isn't passed in here though, the window is stored in the RenderCtx so the resolution is retrieved via a method on that.

Unrelated note about this same code location: `recreate_swapchain` is done both here as well as in response to the out of date error returns during rendering.  On Windows, either of these is sufficient but this one takes precedence and produces nicer looking results (no black bars around the rendered portion on the frame drawn during a resize).

**Comment by Hentropy** [comment]
@Jasper-Bekkers Thanks for the review!
I've fixed the validation errors, resizing works properly now, and the multithreading is no longer racy af.  Hopefully, I've fixed the Linux crash as well but I don't have a Linux machine to test that.

**Comment by khyperia** [comment]
Linux now works (as in doesn't crash, resizes properly), although there's still some validation errors when resizing the window in floating mode (both of these messages get spammed, here's two examples):

```
ERROR:
VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (2094043421)] : Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x55ecdb5aafb8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (746,553), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (745,553), minImageExtent = (745,553), maxImageExtent = (745,553). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)

ERROR:
VALIDATION [VUID-VkFramebufferCreateInfo-pAttachments-00882 (-2119532822)] : Validation Error: [ VUID-VkFramebufferCreateInfo-pAttachments-00882 ] Object 0: handle = 0x55ecdb5aafb8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x81aa7eea | vkCreateFramebuffer(): VkFramebufferCreateInfo attachment #0 mip level 0 has dimensions smaller than the corresponding framebuffer dimensions. Here are the respective dimensions for attachment #0, framebuffer:
width: 742, 743
height: 549, 549
layerCount: 1, 1
```

**Comment by Hentropy** [comment]
@khyperia Sweet! Those validation errors should be cleared up now too.

**Comment by khyperia** [comment]
Hmm, looks like the second one got fixed, but the first one is still there.

```
ERROR:
VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (2094043421)] : Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x55ffe0fc9fc8, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (194,178), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (196,178), minImageExtent = (196,178), maxImageExtent = (196,178). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
```

There's no startup or shutdown warnings, though, which is hecking great!

**Comment by Hentropy** [comment]
I'm not sure how to fix that last linux validation error.  From what I can tell without a linux machine, the OS is resizing the window between the call to `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` right before creating the swapchain and the second call to `vkGetPhysicalDeviceSurfaceCapabilitiesKHR` in the validation layer.

**Comment by Hentropy** [comment]
Those last two changes address an issue with wgpu on laptops with iGPU and dGPU.  Users will still have to set their default 3d GPU to their dGPU though, or they will get a confusing error: "error X3000: unrecognized identifier `shared_ShaderConstants`" which is coming from DX12.
This is ready for a second review.  Linux and macOS should both get checked again, though that one validation error will still be there.

**Review by Jasper-Bekkers — APPROVED** [review]
With the new changes it looks like I'm no longer getting validation messages on Windows, one of the discord members (DeanBDean) tested on macOS as well.

Like @khyperia mentioned, it's probably useful from a pedagogical point of view to have this just rely on spirv-builder instead of doing it's own thing since we expect our users to go through spirv-builder as well, but that can be done in a follow up PR.

**Review by khyperia — APPROVED** [review]

**Pull request merged**
